runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java

```diff
@@ -103,60 +116,67 @@ protected ILogData getNextEntry(QueuedStreamContext context,
     @Override
     protected List<ILogData> getNextEntries(QueuedStreamContext context, long maxGlobal,
                                             Function<ILogData, Boolean> contextCheckFn) {
+        NavigableSet<Long> readSet = new TreeSet<>();
+
+        // Scan backward in the stream to find interesting
+        // log records less than or equal to maxGlobal.
+        // Boolean includes both CHECKPOINT & DATA entries.
+        boolean readQueueIsEmpty = !fillReadQueue(maxGlobal, context);
+
+        // If we witnessed a checkpoint during our scan that
+        // we should pay attention to, then start with them.
+        readSet.addAll(context.readCpQueue);
+
         // We always have to fill to the read queue to ensure we read up to
         // max global.
-        if (!fillReadQueue(maxGlobal, context)) {
+        if (readQueueIsEmpty) {
             return Collections.emptyList();
         }
 
-        // If the lowest element is greater than maxGlobal, there's nothing
-        // to return.
-        if (context.readQueue.first() > maxGlobal) {
-            return Collections.emptyList();
+        if (!context.readQueue.isEmpty() && context.readQueue.first() > maxGlobal) {
```

Comment:
`readSet` may contain log addresses for checkpoint data.  I'll move the `if (readQueueIsEmpty)` statement a bit earlier to help clarify.